Speed up particle sorting.

gassmoeller opened the pull request:
These additional lines of code improve the scaling for large number of cells and particles per subdomain, and can save up to 75 % of the total particle computation cost in a few cases. Realistic speed gains are around 10 %.

The sorting into new cells now first considers neighbors in the order of the distance between particle and the center of the face that connects the old cell with the neighbor.

bangerth commented on source/particle/world.cc:
```diff
@@ -501,6 +501,53 @@ namespace aspect
+          if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+            return true;
+        }
+      catch (...)
```
Can you try to only catch the one specific exception type that indicates that the point was not found? All other exceptions are then allowed to propagate.

bangerth commented on source/particle/world.cc:
```diff
@@ -501,6 +501,53 @@ namespace aspect
+    World<dim>::neighbor_cells_to_search(const Particle<dim> &particle,
+                                         const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const
+    {
+      std::map<double,typename parallel::distributed::Triangulation<dim>::active_cell_iterator> neighbor_cells;
```
I can imagine simple situations where multiple cells haopen to have the same distance. Maybe use a multimap?

bangerth replied:
(This will happen rarely and we still get the same result, but we may as well treat it right. )

bangerth commented on source/particle/world.cc:
```diff
@@ -501,6 +501,53 @@ namespace aspect
+              // be connected to our original cell, because level
+              // differences across faces are limited to 1.
+              for (unsigned int child_index=0; child_index<cell->neighbor(face_no)->n_children(); ++child_index)
+                if (cell->neighbor(face_no)->child(child_index)->active())
```
I don't know if it's worth it, but you could use `cell->neighbor_child_on_subface` to only get the immediate neighbors.

bangerth commented on source/particle/world.cc:
```diff
@@ -522,58 +569,76 @@ namespace aspect
         {
+          // The cell the particle is in
+          typename parallel::distributed::Triangulation<dim>::active_cell_iterator current_cell;
+          bool found_cell(false);
```
Not wrong but we usually initialize scalar variables with `= false`.

bangerth commented:
Nice improvement! I just have a few small niggles.

I could think of a further improvement by sorting cells not by proximity to the particle but by taking into account the particle's direction of movement (I. E. Velocity). But I don't know if you have this available.

bangerth commented:
(And I'd be quite happy to merge the current version plus minor edits if you prefer. )

gassmoeller commented:
I think I implemented all your comments except the evaluation of the velocity. I would like to postpone that test, because I do not know if evaluating the velocity at the cell center, or particle position would really be faster than testing one or two additional cells. I would need to set up a benchmark for that.
Could you take a look if I implemented the loop over the neighbor children correctly? This neighbor_of_neighbor function is new to me.

bangerth commented:
Yes, this looks correct to me. Please rebase, add a changelog entry, and merge!
